packages/manager/apps/account-disable-2fa/src/utils/invisible-challenge.ts

```diff
@@ -0,0 +1,165 @@
+import { AxiosInstance, InternalAxiosRequestConfig } from 'axios';
+import { IframeLoaderPayload, loadIframe, removeIframe } from '@/utils/iframe';
+
+export type ChallengeApiInterceptorChallengeResponse = {
+  challenge_response: string;
+};
+
+type Deferred<T> = {
+  promise: Promise<T>;
+  resolve?: (value: T) => void;
+  reject?: (reason?: ChallengeFailureReasons) => void;
+};
+
+export enum ChallengeFailureReasons {
+  NoIframe,
+  Timeout,
+}
+
+export const CHALLENGE_IFRAME_ID = 'ovh_security_challenge';
+export const TTL_TO_WAIT_IFRAME_RESPONSE = 1000;
+export const INVISIBLE_CHALLENGE_ERROR_CLASS =
+  'Client::BadRequest::ChallengeRequired';
+export const IFRAME_STYLE = 'position: absolute; width:0; height:0; border:0;';
+
+export class ChallengeApiInterceptor {
+  private apiClient: AxiosInstance;
+
+  private iframe: HTMLIFrameElement | null;
+
+  private deferWaitingChallenge: Deferred<unknown> | null;
+
+  private abortTimeout: number;
+
+  constructor(apiClient: AxiosInstance) {
+    this.apiClient = apiClient;
+    this.iframe = null;
+    this.deferWaitingChallenge = null;
+    this.abortTimeout = null;
+  }
+
+  public attachMessageListener() {
+    window.addEventListener(
+      'message',
+      (message: MessageEvent) => this.onIframeMessage(message),
+      false,
+    );
+  }
+
+  public loadChallenge(url: string): Promise<unknown> {
+    if (this.deferWaitingChallenge) {
+      return this.deferWaitingChallenge.promise;
+    }
+
+    // alternative to Promise.withResolvers()
+    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers
+    this.deferWaitingChallenge = {
+      promise: null,
+      resolve: () => null,
+      reject: () => null,
+    };
+    this.deferWaitingChallenge.promise = new Promise((resolve, reject) => {
+      this.deferWaitingChallenge.resolve = resolve;
+      this.deferWaitingChallenge.reject = reject;
+    });
+
+    // If the iframe is already loaded
+    if (this.iframe) {
+      // And the payload has not changed, we can re-use it.
+      if (this.iframe.src === url) {
+        // This iframe is already loaded, go forward to talk with it
+        this.onIframeLoaded();
+        return this.deferWaitingChallenge.promise;
+      }
+      // Otherwise we remove the previous one to use a new one
+      removeIframe(this.iframe);
+      this.iframe = null;
+    }
+    const payload: IframeLoaderPayload = {
+      src: url,
+      id: CHALLENGE_IFRAME_ID,
```

Comment:
**nitpick:** Is it possible to load more than 1 challenge at the same time? If yes, then we will have more than 1 frame with same IDs otherwise ignore my comment.

Reply:
We can't have multiple iframe with the same ID as the code right before this handle the case of an iframe already present (and remove it if the url of the challenge isn't the same)
Moreover, in the case of 2fa, we should not have concurrent requests and as a result only one challenge request at a time (unless we incorrectly implement the UI and let the user click multiple times on "Send documents")